In [12]:
%run "utilspro.py"


In [13]:
# set truncation threshold to display all rows
pd.set_option('display.max_columns', None)
# Loading the dataset
data = pd.read_csv(dataset_path)

# Displaying the first few rows of the dataset
data.head()

number incident_state  active  reassignment_count  reopen_count  \
0  INC0000045            New    True                   0             0   
1  INC0000045       Resolved    True                   0             0   
2  INC0000045       Resolved    True                   0             0   
3  INC0000045         Closed   False                   0             0   
4  INC0000047            New    True                   0             0   

   sys_mod_count  made_sla    caller_id       opened_by        opened_at  \
0              0      True  Caller 2403    Opened by  8  29/2/2016 01:16   
1              2      True  Caller 2403    Opened by  8  29/2/2016 01:16   
2              3      True  Caller 2403    Opened by  8  29/2/2016 01:16   
3              4      True  Caller 2403    Opened by  8  29/2/2016 01:16   
4              0      True  Caller 2403  Opened by  397  29/2/2016 04:40   

   sys_created_by   sys_created_at  sys_updated_by   sys_updated_at  \
0    Created by 6  29/2/2016 01:23   Updated by 21  29/2/2016 01:23   
1    Created by 6  29/2/2016 01:23  Updated by 642  29/2/2016 08:53   
2    Created by 6  29/2/2016 01:23  Updated by 804  29/2/2016 11:29   
3    Created by 6  29/2/2016 01:23  Updated by 908   5/3/2016 12:00   
4  Created by 171  29/2/2016 04:57  Updated by 746  29/2/2016 04:57   

  contact_type      location     category      subcategory    u_symptom  \
0        Phone  Location 143  Category 55  Subcategory 170   Symptom 72   
1        Phone  Location 143  Category 55  Subcategory 170   Symptom 72   
2        Phone  Location 143  Category 55  Subcategory 170   Symptom 72   
3        Phone  Location 143  Category 55  Subcategory 170   Symptom 72   
4        Phone  Location 165  Category 40  Subcategory 215  Symptom 471   

  cmdb_ci      impact     urgency      priority assignment_group  assigned_to  \
0       ?  2 - Medium  2 - Medium  3 - Moderate         Group 56            ?   
1       ?  2 - Medium  2 - Medium  3 - Moderate         Group 56            ?   
2       ?  2 - Medium  2 - Medium  3 - Moderate         Group 56            ?   
3       ?  2 - Medium  2 - Medium  3 - Moderate         Group 56            ?   
4       ?  2 - Medium  2 - Medium  3 - Moderate         Group 70  Resolver 89   

   knowledge  u_priority_confirmation         notify problem_id rfc vendor  \
0       True                    False  Do Not Notify          ?   ?      ?   
1       True                    False  Do Not Notify          ?   ?      ?   
2       True                    False  Do Not Notify          ?   ?      ?   
3       True                    False  Do Not Notify          ?   ?      ?   
4       True                    False  Do Not Notify          ?   ?      ?   

  caused_by closed_code      resolved_by      resolved_at       closed_at  
0         ?      code 5  Resolved by 149  29/2/2016 11:29  5/3/2016 12:00  
1         ?      code 5  Resolved by 149  29/2/2016 11:29  5/3/2016 12:00  
2         ?      code 5  Resolved by 149  29/2/2016 11:29  5/3/2016 12:00  
3         ?      code 5  Resolved by 149  29/2/2016 11:29  5/3/2016 12:00  
4         ?      code 5   Resolved by 81   1/3/2016 09:52  6/3/2016 10:00

In [14]:
# diplay the info of the dataset and the number of rows and columns and the nulls per columns in same functio
# Extracting modes
modes = data.mode()

# Helper function to get the three most common
def three_most_common(lst):
    count = Counter(lst)
    most_common = count.most_common(3)
    return [item[0] for item in most_common]



# Creating an information DataFrame
info_df = pd.DataFrame({
    'Column': data.columns,
    'Dtype': data.dtypes.values,
    'Null Count': data.isnull().sum().values,
    'Non-Null Count': data.count().values,
    'Unique Count': data.nunique().values,
    '1st Mode': [three_most_common(data[col])[0] if len(three_most_common(data[col])) > 0 else None for col in data.columns],
    '2nd Mode': [three_most_common(data[col])[1] if len(three_most_common(data[col])) > 1 else None for col in data.columns],
    '3rd Mode': [three_most_common(data[col])[2] if len(three_most_common(data[col])) > 2 else None for col in data.columns]
})

info_df


Column   Dtype  Null Count  Non-Null Count  Unique Count  \
0                    number  object           0          141712         24918   
1            incident_state  object           0          141712             9   
2                    active    bool           0          141712             2   
3        reassignment_count   int64           0          141712            28   
4              reopen_count   int64           0          141712             9   
5             sys_mod_count   int64           0          141712           115   
6                  made_sla    bool           0          141712             2   
7                 caller_id  object           0          141712          5245   
8                 opened_by  object           0          141712           208   
9                 opened_at  object           0          141712         19849   
10           sys_created_by  object           0          141712           186   
11           sys_created_at  object           0          141712         11553   
12           sys_updated_by  object           0          141712           846   
13           sys_updated_at  object           0          141712         50664   
14             contact_type  object           0          141712             5   
15                 location  object           0          141712           225   
16                 category  object           0          141712            59   
17              subcategory  object           0          141712           255   
18                u_symptom  object           0          141712           526   
19                  cmdb_ci  object           0          141712            51   
20                   impact  object           0          141712             3   
21                  urgency  object           0          141712             3   
22                 priority  object           0          141712             4   
23         assignment_group  object           0          141712            79   
24              assigned_to  object           0          141712           235   
25                knowledge    bool           0          141712             2   
26  u_priority_confirmation    bool           0          141712             2   
27                   notify  object           0          141712             2   
28               problem_id  object           0          141712           253   
29                      rfc  object           0          141712           182   
30                   vendor  object           0          141712             5   
31                caused_by  object           0          141712             4   
32              closed_code  object           0          141712            18   
33              resolved_by  object           0          141712           217   
34              resolved_at  object           0          141712         18506   
35                closed_at  object           0          141712          2707   

           1st Mode         2nd Mode         3rd Mode  
0        INC0019396       INC0044260       INC0005927  
1            Active              New         Resolved  
2              True            False             None  
3                 0                1                2  
4                 0                1                2  
5                 0                1                2  
6              True            False             None  
7       Caller 1904       Caller 290      Caller 4514  
8     Opened by  17    Opened by  24   Opened by  131  
9   14/4/2016 20:42   4/7/2016 14:38   7/3/2016 09:11  
10                ?    Created by 10    Created by 13  
11                ?   4/7/2016 14:50  15/3/2016 15:19  
12   Updated by 908    Updated by 44    Updated by 60  
13  24/3/2016 18:40  24/3/2016 19:00  24/3/2016 18:51  
14            Phone     Self service            Email  
15     Location 204     Location 161     Location 143  
16      Category 26      Category 42      Category 53  
17  Subcategory

In [8]:
subdata

comments:


## Data structure and summary

In [ ]:
# Checking the data structure (data types and missing values)
data_info = data.info()

# Checking for placeholder values (assuming '?' is a placeholder)
placeholder_counts = data.apply(lambda x: x[x == '?'].count())

# Getting statistical summary of the dataset for numerical columns
data_summary = data.describe()

data_info, placeholder_counts, data_summary


* #### Data Columns and Types:
The dataset has 141,712 entries and 36 columns.
Data types include objects (strings), integers, and booleans.
Several columns like opened_at, resolved_at, and closed_at are of object type (strings) representing dates. We may need to convert these to datetime format for further analysis.

* #### Placeholder Values (?):
Several columns contain the placeholder value ?. For instance, the cmdb_ci column has 141,267 such values, which is a significant portion of the total entries.
Other columns with a high count of placeholders include sys_created_by, u_symptom, problem_id, rfc, vendor, and caused_by.

* ### Statistical Summary:
reassignment_count ranges from 0 to 27, with an average of around 1.1.
reopen_count has a maximum value of 8, but 75% of the data is still 0, indicating that most incidents are not reopened.
sys_mod_count (probably indicating system modifications) has a wide range, with values from 0 to 129 and an average of around 5.

#### Granularity and uniqueness


In [ ]:
# Checking the uniqueness of the 'number' column
unique_incidents = data['number'].nunique()
total_entries = len(data)

unique_incidents, total_entries

* 24,918 unique incidents (as indicated by the number column).
* 141,712 total entries.

This suggests that the dataset is not at the granularity of individual incidents. Instead, each incident has multiple entries, possibly capturing different states or updates related to the incident over time.

### Data distribution

We'll focus on the following columns:

 * Numerical: reassignment_count, reopen_count, and sys_mod_count.
* Categorical: incident_state, contact_type, and priority.

In [ ]:
# Setting up the plotting environment
sns.set_style("whitegrid")
plt.figure(figsize=(15, 10))

# Plotting the distribution of the numerical columns
numerical_columns = ['reassignment_count', 'reopen_count', 'sys_mod_count']

for i, column in enumerate(numerical_columns, 1):
    plt.subplot(2, 3, i)
    sns.histplot(data[column], kde=True, bins=30)
    plt.title(f'Distribution of {column}')
    plt.xlabel(column)
    plt.ylabel('Frequency')

plt.tight_layout()
plt.show()

* ### Reassignment Count:
 Most incidents are reassigned once or not at all, with a sharp decline in frequency as the number of reassignments increases.
* ### Reopen Count:
The vast majority of incidents are not reopened. Only a small number of incidents have been reopened multiple times.
* ### Sys Mod Count (System Modification Count):
 Most incidents have a system modification count between 0 and 10. However, there's a long tail, indicating that some incidents have been modified many time

 let's visualize the distribution of the selected categorical columns: incident_state, contact_type, and priority.

In [ ]:
# Plotting the distribution of the categorical columns
categorical_columns = ['incident_state', 'contact_type', 'priority']

plt.figure(figsize=(15, 10))

for i, column in enumerate(categorical_columns, 1):
    plt.subplot(2, 3, i)
    sns.countplot(data=data, y=column, order=data[column].value_counts().index)
    plt.title(f'Distribution of {column}')
    plt.xlabel('Frequency')
    plt.ylabel(column)

plt.tight_layout()
plt.show()

* ### Incident State:
The most common state is "Active", followed by "New" and "Resolved".
States like "Awaiting Problem", "Awaiting Vendor", and "Awaiting Evidence" have considerably fewer occurrences.
* ### Contact Type:
The vast majority of incidents are reported via "Phone", with very few incidents reported through "Email", "Self service", or "Direct opening".
* ### Priority:
The "3 - Moderate" priority level is the most common, followed by "2 - High" and "4 - Low". There are fewer incidents with "1 - Critical" and "5 - Very Low" priorities.

### Correlation analysis

In [ ]:
# Correlation matrix for numerical columns
correlation_matrix = data[numerical_columns].corr()

# Plotting the correlation matrix using a heatmap
plt.figure(figsize=(10, 7))
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', linewidths=0.5, vmin=-1, vmax=1)
plt.title('Correlation Matrix of Numerical Variables')
plt.show()

* ### Reassignment Count and Sys Mod Count: 
There's a light positive correlation (approximately 0.53) between these two variables. This indicates that as incidents are reassigned more often, the number of system modifications also tends to increase.
* ### Reopen Count:
This variable doesn't have a strong correlation with the other two. Its correlation with both reassignment_count and sys_mod_count is relatively low.

### Anomaly Detection

We'll focus on identifying potential outliers within the numerical columns. Outliers can distort the results of our analyses and models.

In [ ]:
# Plotting boxplots for anomaly detection in numerical columns
plt.figure(figsize=(15, 7))

for i, column in enumerate(numerical_columns, 1):
    plt.subplot(1, 3, i)
    sns.boxplot(data[column])
    plt.title(f'Box Plot of {column}')
    plt.xlabel(column)

plt.tight_layout()
plt.show()

* #### Reassignment Count:
Most of the data points lie between 0 and 2. However, there are several outliers beyond this range, with some incidents having been reassigned more than 20 times.
* #### Reopen Count:
While the majority of incidents are not reopened, there are outliers where incidents have been reopened multiple times.
* #### Sys Mod Count:
Most of the incidents have undergone system modifications less than 15 times. Yet, there are outliers, with some incidents having more than 40 system modifications.

## Data Consistency 
we'll focus on checking if:
* There are incidents with a closed_at date earlier than the opened_at date.
* Identifying any incidents that are marked as active but have a closed_at date.

In [ ]:
# Converting 'opened_at', 'closed_at' to datetime format
data['opened_at'] = pd.to_datetime(data['opened_at'], errors='coerce', dayfirst=True)
data['closed_at'] = pd.to_datetime(data['closed_at'], errors='coerce', dayfirst=True)

# Checking for incidents with 'closed_at' date earlier than 'opened_at' date
inconsistent_dates = data[data['closed_at'] < data['opened_at']]

# Checking for incidents that are marked as 'active' but have a 'closed_at' date
active_but_closed = data[(data['active'] == True) & (~data['closed_at'].isna())]

inconsistent_dates_count = len(inconsistent_dates)
active_but_closed_count = len(active_but_closed)

inconsistent_dates_count, active_but_closed_count

* #### There are no incidents with a closed_at date earlier than the opened_at date.
 This is good as it indicates consistent date information.
* #### There are 116,726 incidents that are marked as active but have a closed_at date.
 This is peculiar and suggests potential inconsistencies in the data. Incidents that are still active shouldn't have a closure date.

NB :  It's crucial to address identified inconsistencies and anomalies before proceeding with  building machine learning models.